File: java/revenj-servlet/src/main/java/org/revenj/server/servlet/RevenjSerialization.java
package org.revenj.server.servlet;

import com.dslplatform.json.DslJson;
import com.dslplatform.json.JsonWriter;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.revenj.serialization.json.DslJsonSerialization;
import org.revenj.serialization.Serialization;
import org.revenj.patterns.ServiceLocator;
import org.revenj.serialization.WireSerialization;
import org.revenj.serialization.xml.XmlJaxbSerialization;
import org.w3c.dom.Element;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.reflect.Type;
import java.util.Optional;

final class RevenjSerialization implements WireSerialization {
	private final DslJsonSerialization json;
	private final XmlJaxbSerialization xml;
	private final PassThroughSerialization passThrough;

	public RevenjSerialization(ServiceLocator locator, XmlJaxbSerialization xml) {
		JacksonSerialization jackson = new JacksonSerialization(locator, locator.tryResolve(ObjectMapper.class));
		this.json = new DslJsonSerialization(locator, Optional.of(new DslJson.Fallback<ServiceLocator>() {
			@Override
			public void serialize(Object instance, OutputStream stream) throws IOException {
				jackson.serialize(instance, stream);
			}

			@Override
			public Object deserialize(ServiceLocator serviceLocator, Type manifest, byte[] body, int size) throws IOException {
				return jackson.deserialize(manifest, body, size);
			}

			@Override
			public Object deserialize(ServiceLocator serviceLocator, Type manifest, InputStream stream) throws IOException {
				return jackson.deserialize(manifest, stream);
			}
		}));
		this.passThrough = new PassThroughSerialization();
		this.xml = xml;
	}

	private static final ThreadLocal<JsonWriter> threadWriter = new ThreadLocal<JsonWriter>() {
		@Override
		protected JsonWriter initialValue() {
			return new JsonWriter();
		}
	};
	private static final ThreadLocal<byte[]> threadBuffer = new ThreadLocal<byte[]>() {
		@Override
		protected byte[] initialValue() {
			return new byte[65536];
		}
	};

	@Override
	public String serialize(Object value, OutputStream stream, String accept) throws IOException {
		if (accept != null && accept.startsWith("application/xml")) {
			xml.serializeTo(value, stream);
			return "application/xml; charset=UTF-8";
		}
		JsonWriter writer = threadWriter.get();
		writer.reset();
		json.serialize(writer, value);
		writer.toStream(stream);
		return "application/json";
	}

	@Override
	public Object deserialize(Type type, byte[] content, int length, String contentType) throws IOException {
		if (contentType != null && contentType.startsWith("application/xml")) {
			ByteArrayInputStream is = new ByteArrayInputStream(content, 0, length);
			return xml.deserialize(type, is);
		}
		return json.deserialize(type, content, length);
	}

	@Override
	public Object deserialize(Type type, InputStream stream, String contentType) throws IOException {
		if (contentType != null && contentType.startsWith("application/xml")) {
			return xml.deserialize(type, stream);
		}
		return json.deserialize(type, stream, threadBuffer.get());
	}

	@Override
	public <TFormat> Optional<Serialization<TFormat>> find(Class<TFormat> format) {
		if (Object.class.equals(format)) {
			return Optional.of((Serialization) passThrough);
		} else if (String.class.equals(format)) {
			return Optional.of((Serialization) json);
		} else if (Element.class.equals(format)) {
			return Optional.of((Serialization) xml);
		}
		return Optional.empty();
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/RevenjSerialization.scala
package net.revenj

import java.io.{ByteArrayInputStream, ByteArrayOutputStream, InputStream, OutputStream}
import java.lang.reflect.Type
import java.nio.charset.StandardCharsets
import net.revenj.RevenjSerialization.PassThroughSerialization
import net.revenj.serialization.{DslJsonSerialization, JacksonSerialization, Serialization, WireSerialization}

import java.util.Properties
import scala.reflect.runtime.universe._
import scala.util.{Failure, Success, Try}

private[revenj] class RevenjSerialization(
  dslJson: DslJsonSerialization,
  jackson: JacksonSerialization,
  properties: Properties,
  loader: ClassLoader
) extends WireSerialization {
  private val passThrough = new PassThroughSerialization()
  private val mirror = runtimeMirror(loader)

  private val useDslJson = properties.getProperty("revenj.serialization") == "dsl-json"

  import RevenjSerialization._

  private val sharedBuffer = new ThreadLocal[Array[Byte]] {
    override def initialValue(): Array[Byte] = new Array[Byte](4096)
  }

  private val arrayFormat = mirror.typeOf[Array[Byte]]
  private val streamFormat = mirror.typeOf[InputStream]

  private def inputToOutput(input: InputStream, output: OutputStream): Unit = {
    val buffer = sharedBuffer.get()
    var read = 0
    while ({read = input.read(buffer); read != -1}) {
      output.write(buffer, 0, read)
    }
  }

  override def serialize(value: Any, stream: OutputStream, contentType: String, manifest: Type): Try[String] = {
    value match {
      case byte: Array[Byte] if isBinary(contentType) =>
        stream.write(byte)
        Success(contentType)
      case is: InputStream =>
        inputToOutput(is, stream)
        Success(contentType)
      case opt: Option[_] if opt.isDefined => serialize(opt.get, stream, contentType, manifest)
      case _ =>
        require(contentType == "application/json", "Only application/json content type is supported")
        if (value != null) {
          if (useDslJson) {
            dslJson.serializeRuntime(value, stream, manifest).map(_ => "application/json")
          } else {
            jackson.serialize(value, stream, manifest).map(_ => "application/json")
          }
        } else {
          Try(stream.write(Null)).map(_ => "application/json")
        }
    }
  }

  override def deserialize(manifest: Type, content: Array[Byte], length: Int, contentType: String): Try[Any] = {
    if (isBinary(contentType)) {
      Success(content)
    } else {
      require(contentType == "application/json", "Only application/json content type is supported")
      if (useDslJson) {
        dslJson.deserializeRuntime(content, length, manifest)
      } else {
        jackson.deserialize(manifest, content, length)
      }
    }
  }

  override def deserialize(manifest: Type, stream: InputStream, contentType: String): Try[Any] = {
    if (isBinary(contentType)) {
      Success(stream)
    } else {
      require(contentType == "application/json", "Only application/json content type is supported")
      if (useDslJson) {
        dslJson.deserializeRuntime(stream, manifest)
      } else {
        jackson.deserialize(manifest, stream)
      }
    }
  }

  override def deserialize[T: TypeTag](content: Array[Byte], length: Int, contentType: String): Try[T] = {
    val format = mirror.typeOf[T]
    Utils.findType(format, mirror) match {
      case Some(tpe) =>
        if (isBinary(contentType)) {
          if (format == arrayFormat) {
            Success(content.asInstanceOf[T])
          } else if (format == streamFormat) {
            Success(new ByteArrayInputStream(content).asInstanceOf[T])
          } else {
            throw new IllegalArgumentException("Invalid format specified. Only Array[Byte] and InputStream supported")
          }
        } else {
          require(contentType == "application/json", "Only application/json content type is supported")
          deserialize(tpe, content, length, contentType) match {
            case Success(value: T@unchecked) => Success(value)
            case Failure(ex) => Failure(ex)
          }
        }
      case _ =>
        Failure(new IllegalArgumentException(s"Unable to find Java type for: ${mirror.typeOf[T]}"))
    }
  }

  override def deserialize[T: TypeTag](stream: InputStream, contentType: String): Try[T] = {
    val format = mirror.typeOf[T]
    Utils.findType(format, mirror) match {
      case Some(tpe) =>
        if (isBinary(contentType)) {
          if (format == arrayFormat) {
            val baos = new ByteArrayOutputStream()
            inputToOutput(stream, baos)
            Success(baos.toByteArray.asInstanceOf[T])
          } else if (format == streamFormat) {
            Success(stream.asInstanceOf[T])
          } else {
            throw new IllegalArgumentException("Invalid format specified. Only Array[Byte] and InputStream supported")
          }
        } else {
          require(contentType == "application/json", "Only application/json content type is supported")
          deserialize(tpe, stream, contentType) match {
            case Success(value: T@unchecked) => Success(value)
            case Failure(ex) => Failure(ex)
          }
        }
      case _ =>
        Failure(new IllegalArgumentException(s"Unable to find Java type for: ${mirror.typeOf[T]}"))
    }
  }

  override def find[TFormat: TypeTag](): Option[Serialization[TFormat]] = {
    val format = mirror.typeOf[TFormat]
    if (typeOf[Any] == format || typeOf[AnyRef] == format) {
      Some(passThrough.asInstanceOf[Serialization[TFormat]])
    } else if (typeOf[String] == format) {
      if (useDslJson) Some(dslJson.asInstanceOf[Serialization[TFormat]])
      else Some(jackson.asInstanceOf[Serialization[TFormat]])
    } else {
      None
    }
  }
}
private object RevenjSerialization {
  private val Null = "null".getBytes(StandardCharsets.UTF_8)

  private def isBinary(contentType: String): Boolean = {
    contentType.startsWith("image/") || contentType.startsWith("vide/") || Binary.contains(contentType)
  }

  private val Binary = Set(
    "application/octet-stream",
    "application/pdf",
    "application/msword",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "application/vnd.ms-word.document.macroEnabled.12",
    "application/vnd.ms-excel",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "application/vnd.ms-excel.sheet.macroEnabled.12",
    "application/vnd.ms-powerpoint",
    "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    "application/vnd.ms-powerpoint.presentation.macroEnabled.12"
  )

  private class PassThroughSerialization extends Serialization[Any] {
    override def serialize[T: TypeTag](value: T): Try[Any] = Success(value)

    override private[revenj] def serializeRuntime(value: Any, manifest: Type): Try[Any] = Success(value)

    override def deserialize[T: TypeTag](input: Any): Try[T] = Success(input.asInstanceOf[T])

    override private[revenj] def deserializeRuntime[T](input: Any, manifest: Type): Try[T] = Success(input.asInstanceOf[T])
  }
}